### DISCUSSION TITLE
Unable to access XML nodes by name in a code template when strict parsing is enabled

### DISCUSSION THREAD

**Discussion by GrantMilesNHS** [nextgenhealthcare/connect#5349] [Q&A] [answered]
**Describe the bug**

Many of our channels extract the patient ID for custom metadata columns and destination mappings. We have recently attempted to move the repeated bit of code which retrieves the id from each source transformer and put it into a code library whereby each source transformer can call and then store in the channel map. However once moving the code (show below) into a code template we are no longer able to get the PID section from the msg object.

```js
var ourid = "";
for (var i = 0; i < getArrayOrXmlLength(hl7_msg['PID']['PID.3']); i++) {	
	var assigningAuthority = hl7_msg['PID'][i]['PID.3']['CX.4']['HD.1'].toString()
	if (assigningAuthority == "MRN")
	{
		try{
			ourid = hl7_msg['PID'][i]['PID.3']['CX.1'].toString()
		} catch (e) {
			//logger.error(logPrefix + 'Error : ' + e.message);
			ourid = "";
		}
	}
}
logger.info('extracted Ourid : ' + ourid);
return ourid;
```

I believe the code was taken from the generated JS when using the drag and drop functionality in mirth connect and works fine when run inside the source transformer. You are able to navigate the the XML nodes if you use the `.children()[0]` notation however this is unfeasible as a long term solution.  

**To Reproduce**

Create a new channel and set data types to be HL7 v2 with strict parsing checked and save

![image](https://user-images.githubusercontent.com/78217276/184096553-f640ce56-9338-4d58-96e5-785db1bfdbc1.png)

Add a new code library called Test Library and give it access to the new channel and paste the following code to retrieve the patient ID
```js
function GO() {
	var ourid = "";
	for (var i = 0; i < getArrayOrXmlLength(msg['PID']['PID.3']); i++) {	
		var assigningAuthority = pid[i]['PID.3']['CX.4']['HD.1'].toString()
		if (assigningAuthority == "MRN")
		{
			try{
				ourid = pid[i]['PID.3']['CX.1'].toString()
			} catch (e) {
				//logger.error(logPrefix + 'Error : ' + e.message);
				ourid = "";
			}
		}
	}
	logger.info('extracted OurId : ' + ourid);
	return ourid;
}
```

Go back into the test channel and add the following into the source transformer
```js
var ourIdAt =  GO();
logger.info('(CHANNEL NAME) SOURCE TRANSFORMER - Adding OurIdAlt = ' + ourIdAt +  ' to channel map');
channelMap.put("ourIdAt", ourIdAt);
```

In the channel writer destination add the ourIdAt varible so you can see the output.

Send the following message to test the id extraction
```csv
MSH|^~\&|SYSTEM|IDENTIFIER|SERVER|SERVER|20220809161815||ADT^A01^ADT_A01|4e1d60abebe|P|2.4|||AL|NE|GBR||EN
EVN||20220809161815|||655199537102|20220802080000
PID|1||1389229^^^MRN^CE||NAME3^NAME2^NAME1^^Miss^^L||20080329000000|F|||House, The Street^Locality^Post Town^County^ZZ99^^PRIMARY||0^Mobile|||||||||A||||||||N
PD1|||^^999|G9999998
PV1|1|I|Testward2^Bay1^9^^^^^^Test ward 2\E\nTest Ward 2 Bay 1|BO|||||C9999999^SURNAME^FORNAME^^^Dr|999||||CO||||99|7^^^IDE^CE|||||||||||||||||||||||||20220802080000
AL1|1|MA|Allergy to cow's milk protein|2|ALLERGY1
```

1. Deploy the channel
2. Clear current dashboard server logs (for visibility)
3. Send a HL7 message into the test channel and view the logs/destination.

**Expected behavior**
The patient ID should be mapped to our variable

**Actual behavior**
The PID section is always returned as undefined and you are unable to navigate any of the nodes by name. 

For reference The PID Section (when logging the message) when strictly parsed looks like this:
```xml
<PID xmlns="urn:hl7-org:v2xml">
    <PID.1>1</PID.1>
    <PID.3>
        <CX.1>1389229</CX.1>
        <CX.4>
            <HD.1>MRN</HD.1>
        </CX.4>
        <CX.5>CE</CX.5>
    </PID.3>
    ... Removed for brevity
</PID>
```

**Environment (please complete the following information):**
 - OS: Windows Server 2019
 - Java Distribution/Version openjdk 17.0.3
 - Connect Version 4.0.1

**Workaround(s)**
in the channels data types if you check the "strip namespaces" toggle then you are able to access the nodes by name e.g. `msg['PID']`

**Comment by pacmano1** [comment]
Can you throw a sample hl7 message in your post?

**Comment by pacmano1** [comment]
Just a thought on the code in general:

You might consider using @tonygermano's xFilter function at https://github.com/nextgenhealthcare/connect-examples/tree/master/Code%20Templates/Filter%20XMLLists

Getting that MRN becomes:

```
var ourid = xFilter(msg['PID']['PID.3'], function(pid3) {return pid3['CX.4']['HD.1'] == "MRN"}, 0, 1)
$c('ourid',ourid['CX.1'].toString())
```

**Comment by GrantMilesNHS** [comment]
We are still finding our feet with Mirth Connect, I will take a look at the linked repo tomorrow thank you. I have added a test message and an example of the PID segment to show the transformed message structure

**Comment by GrantMilesNHS** [comment]
> Just a thought on the code in general:
> 
> You might consider using @tonygermano's xFilter function at https://github.com/nextgenhealthcare/connect-examples/tree/master/Code%20Templates/Filter%20XMLLists
> 
> Getting that MRN becomes:
> 
> ```
> var ourid = xFilter(msg['PID']['PID.3'], function(pid3) {return pid3['CX.4']['HD.1'] == "MRN"}, 0, 1)
> $c('ourid',ourid['CX.1'].toString())
> ```

Had a look at this today, it has the same issue as navigating the message manually (when strictly parsed). You need to have the namespaces stripped at which point we have issues on the outbound message serialisation back into HL7.

This is only the case when the code is in a Code Library -> Code Template. When it is included as a transformer step as JavaScript it works fine (either way).

**Comment by DavidBunchNHS** [comment]
I have looked at this with Grant. The xFilter function does work when called directly from the source connector transformer step, even with the namespace stripped. However, if the xFilter function is called from within a code template function, which in turn is called by the transformer step, it only works with the namespace stripped.

**Comment by pacmano1** [comment]
I also reproduced the behavior you describe here.   I would expect the code template to just work.   I've never used the strict parser though.  Intersting problem!   Hopefully some other folks will chime in.

**Comment by tonygermano** [comment, marked as answer]
It shouldn't matter if it's being called from the transformer or a code template, but the namespace most definitely matters. You either need to strip namespaces, access by namespace, or set the default xml namespace before you call the function.

```
default xml namespace = 'urn:hl7-org:v2xml'
```

Changing the default will affect xml that you are writing as well if your outbound message uses xml with a different namespace (or the default blank namespace.)

It should be noted that prior to mirth 3.6 "strip namespaces" was enabled by default on new channels. See https://github.com/nextgenhealthcare/connect/issues/3043.

If you don't strip namespaces or reset the default, you can do this
```js
// with wildcard namespace
var pid3 = msg.*::PID.*::['PID.3']

// with specific namespace
var hl7v2 = msg.namespace()
// or this would work, too
// var hl7v2 = new Namespace('urn:hl7-org:v2xml')
var pid3 = msg.hl7v2::PID.hl7v2::['PID.3']
```

This is yet another way, but it is extremely esoteric. You might not find another example of this on the internet. However, unlike the other javascript solutions, this syntax is valid in any javascript parser, while the others are e4x specific. I'm putting it here just in case it ends up being useful for someone.

```js
var pid3 = msg[new QName("urn:hl7-org:v2xml", 'PID')][new QName("urn:hl7-org:v2xml", 'PID.3')]

// or to make it more javascripty
const h = name => new QName("urn:hl7-org:v2xml", name)
var pid3 = msg[h('PID')][h('PID.3')]
```

**Comment by siddharthsharma9258** [comment]
I think your problem is here.

hl7_msg['PID']['PID.3']


I have passed the entire `msg` object directly to code templates, with or without strict parsing option, and it picks up.

What is hl7_msg?

**Reply by DavidBunchNHS** [reply]
hl7_msg is an argument passed to the function in the code template - it is msg. As msg is global it isn't necessary to do this, but we do it for readability. We have also tried it without passing it as an argument (please see the function GO() in a later post) and, as expected, it didn't make a difference. 

The problem arises when you place the function in a code template and call it from a transformer source connector step. It works correctly if the code is placed in the transformer step.

For example the following code place in the transformer step will work.

//Strict Parsing = true; Strip Namespace = false

	var ourId = "";
	for (var i = 0; i < getArrayOrXmlLength(msg['PID']['PID.3']); i++) {	
		var assigningAuthority = msg['PID'][i]['PID.3']['CX.4']['HD.1'].toString()
		if (assigningAuthority == "MRN")
		{
			try{
				ourId = msg['PID'][i]['PID.3']['CX.1'].toString()
			} catch (e) {
				logger.error ('Error : ' + e.message);
				ourId = "";
			}
		}
	}
	logger.info('Our Id : ' + ourId);




But, if the same code is placed in a code template and called from the transformer step, it doesn't work unless the Strip Namespace inbound property is set to false.

function getOurId() {
	var ourId = "";
	for (var i = 0; i < getArrayOrXmlLength(msg['PID']['PID.3']); i++) {	
		var assigningAuthority = msg['PID'][i]['PID.3']['CX.4']['HD.1'].toString()
		if (assigningAuthority == "MRN")
		{
			try{
				ourId = msg['PID'][i]['PID.3']['CX.1'].toString()
			} catch (e) {
				logger.error ('Error : ' + e.message);
				ourId = "";
			}
		}
	}
	logger.info('Our Id : ' + ourId);
	return ourId;
}

As we don't want to strip the namespace when using strict parsing and code templates, fully qualified namespaces are required when navigating the xml element levels, as per Tony's solution which worked perfectly.
